power_grid_model_c/power_grid_model/include/power_grid_model/main_core/math_state.hpp

```diff
@@ -22,4 +22,43 @@ inline void clear(MathState& math_state) {
     math_state.y_bus_vec_asym.clear();
 }
 
+template <symmetry_tag sym>
+inline void update_y_bus(MathState& math_state, std::vector<MathModelParam<sym>> const& math_model_params) {
+    auto& y_bus_vec = [&math_state]() -> auto& {
```

Comment:
Is this recommended? `auto` `-> auto`?

Reply:
the two `auto`s mean different things and both are needed.

* the second one declares that the lambda returns a reference
* the first one captures the result of the lambda by reference

The lambda allows for side-effects. The construct used here is:

```cpp
struct Foo;
struct Bar { Foo foo; };

Bar bar;
auto& foo_reference = [&bar]() -> auto& { return bar.foo; }();
  ^                      ^   ^      ^               ^       ^
  0                      1   2      3               4       5
```

which has the same result as first declaring only the function and then calling it, but with the additional advantage that the function is not stored, so the compiler will `inline` and maybe even resolve `constexpr`.

0. declares that `foo_reference` is reference (references cannot be forward-declared, so always needs to be obtained from a lambda in complex cases like this)
1. captures `bar` by reference
2. needed because of trailing return type (only way for lambdas to specify return types)
3. declare that return type is a reference (this is the only way to do that for lambdas without using `return *&result` construct, which may be illegal if the captured result is `const`)
4. return the result type
5. immediately call the function so that it doesn't need to be stored in a variable

hope that explains it

Reply:
Clearly a good read but not what I was asking ;)
My question was since you already know `Foo` as return type, why still write `auto` when the type couldn't be inferred by mortal human brains with simple reading?

Reply:
For the ease of discussion:

> ```c++
> struct Foo;
> struct Bar { Foo foo; };
> 
> Bar bar;
> auto& foo_reference = [&bar]() -> auto& { return bar.foo; }();
>   ^                      ^   ^      ^               ^       ^
>   0                      1   2      3               4       5
> ```

```c++
struct Foo;
struct Bar { Foo ofo; };

Bar bar;
auto& foo_reference = [&bar]() -> Foo& { return bar.ofo; }();
```

Reply:
cause I don't know what the return type is in the case of complex operations. In fact, the return type depends on the template parameters (here: `std::vector<YBus<sym>>&`). In this case, I agree that we could still add it, but it also clutters the code with implementation details.

In the more general case, the return type might actually be more complex and not explicitly depend on the template parameters like `Transformer&` vs `ThreeWindingTransformer&`. In that case, the only alternative to provide a trailing return type other than `auto&` would be using `std::conditional_t`, which is already verbose in the case of 2 supported template parameters - let alone multiple

Reply:
Pending approval upon resolution of this conversation.

Reply:
another advantage of using `auto&` is that the caller very often really doesn't care which type it is, as long as it compiles. the best you can get is slightly better type hinting, but only marginally because it involves - once again - templates

Reply:
`auto` it is then